Make COOP+COEP do not imply crossOriginIsolated.

ArthurSonzogni opened the pull request:
The [specification] currently requires [COOP] + [COEP] to give access to
crossOriginIsolated capabilities like SharedArrayBuffer.

Some platforms can't easily support multiple processes (like Android
Webview). Therefore, they can't really support crossOriginIsolated.
However the are no strong reasons for them not to enforce COEP (and
maybe COOP) when their associated headers are present.

It would be great enforcing COEP (and maybe COOP) on all platforms,
desptie the lack of crossOriginIsolated capabilities.

This patch makes the specification to allow (instead of requiring)
platform to set the crossOriginIsolated flag when both COOP and COEP are
used.

Setting crossOriginIsolated becomes platform dependent. In exchange, we
can enforce COEP (and COOP) in a non platform dependent way, without
conflicting with the specification about crossOriginIsolated.

[Bug]: https://github.com/whatwg/html/issues/6060

[specification]:
https://html.spec.whatwg.org/#cross-origin-opener-policies

[COOP]:
https://html.spec.whatwg.org/#cross-origin-opener-policy

[COEP]:
https://html.spec.whatwg.org/#coep


<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://whatpr.org/html/6098/browsers.html" title="Last updated on Jan 4, 2021, 3:42 PM UTC (9c2fd1f)">/browsers.html</a>  ( <a href="https://whatpr.org/html/6098/07a5b3e...9c2fd1f/browsers.html" title="Last updated on Jan 4, 2021, 3:42 PM UTC (9c2fd1f)">diff</a> )
<a href="https://whatpr.org/html/6098/origin.html" title="Last updated on Jan 4, 2021, 3:42 PM UTC (9c2fd1f)">/origin.html</a>  ( <a href="https://whatpr.org/html/6098/07a5b3e...9c2fd1f/origin.html" title="Last updated on Jan 4, 2021, 3:42 PM UTC (9c2fd1f)">diff</a> )
<a href="https://whatpr.org/html/6098/structured-data.html" title="Last updated on Jan 4, 2021, 3:42 PM UTC (9c2fd1f)">/structured-data.html</a>  ( <a href="https://whatpr.org/html/6098/07a5b3e...9c2fd1f/structured-data.html" title="Last updated on Jan 4, 2021, 3:42 PM UTC (9c2fd1f)">diff</a> )
<a href="https://whatpr.org/html/6098/webappapis.html" title="Last updated on Jan 4, 2021, 3:42 PM UTC (9c2fd1f)">/webappapis.html</a>  ( <a href="https://whatpr.org/html/6098/07a5b3e...9c2fd1f/webappapis.html" title="Last updated on Jan 4, 2021, 3:42 PM UTC (9c2fd1f)">diff</a> )
<a href="https://whatpr.org/html/6098/window-object.html" title="Last updated on Jan 4, 2021, 3:42 PM UTC (9c2fd1f)">/window-object.html</a>  ( <a href="https://whatpr.org/html/6098/07a5b3e...9c2fd1f/window-object.html" title="Last updated on Jan 4, 2021, 3:42 PM UTC (9c2fd1f)">diff</a> )
<a href="https://whatpr.org/html/6098/workers.html" title="Last updated on Jan 4, 2021, 3:42 PM UTC (9c2fd1f)">/workers.html</a>  ( <a href="https://whatpr.org/html/6098/07a5b3e...9c2fd1f/workers.html" title="Last updated on Jan 4, 2021, 3:42 PM UTC (9c2fd1f)">diff</a> )

yutakahirano commented:
@annevk @domenic 

I think it's good to make BCG's cross-origin isolated and agent cluster's cross-origin isolated platform independent, and [cross-origin isolated capability](https://html.spec.whatwg.org/multipage/webappapis.html#concept-settings-object-cross-origin-isolated-capability) platform independent. That means even on Android WebView document.domain settings is no-op when COOP: same-origin and COEP: require-corp is specified. What do you think?

ArthurSonzogni commented:
> I think it's good to make BCG's cross-origin isolated and agent cluster's cross-origin isolated platform independent, and cross-origin isolated capability platform independent. That means even on Android WebView document.domain mutation is no-op when COOP: same-origin and COEP: require-corp is specified. What do you think?

I agree it would be a shame not restricting document.domain.

I have been a bit confused by all the different cross-origin-isolated concept. So we have:
1. Browsing context group cross-origin isolated.
2. agent cluster's cross-origin isolated
3. The windows's cross-origin isolated capability.

I guess I should just turn add a platform-specific exception to:
https://html.spec.whatwg.org/multipage/window-object.html#script-settings-for-window-objects:concept-settings-object-cross-origin-isolated-capability
instead.

yutakahirano commented:
My recommendation would be...

- Adding a note to https://html.spec.whatwg.org/multipage/webappapis.html#concept-settings-object-cross-origin-isolated-capability.
  - Something like: Note: this value is platform dependent, because on some platforms it is difficult to ensure security guarantees we want to keep. There are platform independent invariants though.
     - If the surrounding agent cluster's cross-origin isolated is false, then this value is false.
     - If [the corresponding feature policy](https://html.spec.whatwg.org/multipage/infrastructure.html#cross-origin-isolated-feature) doesn't allow to use the feature, then this value is false.

 - Modifying logic at https://html.spec.whatwg.org/multipage/window-object.html#script-settings-for-window-objects and adding a note.
 - Modifying logic at https://html.spec.whatwg.org/multipage/workers.html#worker-processing-model:concept-workerglobalscope-cross-origin-isolated-capability and adding a note.
 - Modifying the corresponding logic at the service worker logic and adding a note (w3c/ServiceWorker#1545).

domenic commented:
I agree with @yutakahirano's suggestions. To give more detail on a particularly tricky part:

> Modifying logic at https://html.spec.whatwg.org/multipage/window-object.html#script-settings-for-window-objects and adding a note.

My suggestion concretely here is that before step 4, you establish a boolean. Something like:

> 4. Let _canBeCOI_ be true if the user agent is willing to give access to COI features to _window_, and false otherwise.

then you can modify the "cross-origin isolated capability" to be something like:

> Return true if all of the following hold:
> - realm's agent cluster's cross-origin-isolated is true
> - canBeCOI is true
> - window's associated Document is allowed to use the "cross-origin-isolated" feature
> Otherwise, return false

Without establishing the boolean ahead of time, if you just wrote something like "return false at the user agent's discretion" in the COIC algorithm, it would be possible per-spec for the value to change over time.

ArthurSonzogni commented:
Thanks Yutaka and Domenic, this is really helpful!

yutakahirano commented:
@ArthurSonzogni , what's the status of this bug?

ArthurSonzogni commented:
> @ArthurSonzogni , what's the status of this bug?

I just need to apply Domenic suggestions and yours. Nothing looks hard, I just need to produce some work ;-)

I was focused on my main project. Enabling COOP+COEP on every platforms is a side project that I wanted to see progress too, but this isn't my main priority and nobody really depends on this. Maybe you do somehow?

I've booked some time Wednesday for fixing this. I will update this thread.

yutakahirano commented:
Thank you!

ArthurSonzogni commented:
> > Modifying logic at https://html.spec.whatwg.org/multipage/window-object.html#script-settings-for-window-objects and adding a note. **~Yutaka**

> My suggestion concretely here is that before step 4, you establish a boolean. Something like:
> (...)
> Without establishing the boolean ahead of time, if you just wrote something like "return false at the user agent's discretion" in the COIC algorithm, it would be possible per-spec for the value to change over time. **~Domenic**

The section linked above is called once for every "window" object being created. So this doesn't really represent a "global" per-agent variable. Maybe there is another section that is called only once, where I can put this variable instead?

domenic commented:
I don't know if it's important for the _canBeCOI_ boolean to be the same across the entire agent. Having it be per-window seems reasonable, since it's just applying an implementation-defined restriction on top of the existing agent cluster-wide cross-origin-isolated situation.

ArthurSonzogni commented:
I made an update. It's a bit unfortunate that "canBeCOI" can't be made more global. I am defining from "Script settings for Window objects", but I also need to use it from the part about workers.

I would be happy to get some feedback from there. I still need to do the suggested change in the w3c/ServiceWorker repository.

domenic commented:
If we wanted it to be more global, we could modify the agent cluster's cross-origin-isolated at the time that gets set (which is roughly at agent cluster creation time). However, I haven't thought through the full implications of that, and am unsure it's a great idea. I remember feeling that the current design you've uploaded is good, when we discussed this 26 days ago.

I'm currently attending BlinkOn so my time for nontrivial reviews like this one is a bit scarce. But as soon as I'm able (probably Friday?) I'll do a review of the latest changes you've made, and also think more deeply about whether it would be a good idea to make it change the agent cluster's cross-origin isolated. Thoughts from @yutakahirano or @annevk would be welcome in the meantime.

annevk commented on source:
```diff
@@ -79602,6 +79602,14 @@ interface <dfn>BarProp</dfn> {
+   <p class="note">Not every platform can easily support the security properties associated with the
+   <code data-x="cross-origin-isolated-feature">cross-origin-isolated</code> feature. Once the
+   decision to support it has been made, it is guaranteed not to change over time.</p>
+   </li>
```
We should use `<a>implementation-defined</a>` here.

ArthurSonzogni replied:
Done.
Note that I used ```<span>implementation-defined<span>``` instead.

annevk reviewed (commented):
I think this approach is fine. If you wanted something more global you would have to change the browsing context group and agent cluster's cross-origin isolated concept to a tri-state I think, but I'm not sure what the observable differences would be. I would hope none.

annevk commented:
Hmm, thinking about it more that tri-state would be a lot cleaner as it more closely matches what an implementation would want to do. It would not process-isolate the agent cluster after all. I think it's worth another look to see how hard that would be.

ArthurSonzogni commented:
> Hmm, thinking about it more that tri-state would be a lot cleaner as it more closely matches what an implementation would want to do. It would not process-isolate the agent cluster after all. I think it's worth another look to see how hard that would be.

That was my feeling since the beginning. I would be happy to do this.

So, I guess I would need to modify:
https://html.spec.whatwg.org/#bcg-cross-origin-isolated

Replace:
```
A browsing context group has a cross-origin isolated boolean. It is initially false.
```
by:
```
A browsing context group has a cross-origin-isolated tri-state. It can take the following values:
- process-isolation
- logical-isolation
- none

It is initially 'none'.
```

Then modify:
https://html.spec.whatwg.org/#browsing-context-group-switches-due-to-cross-origin-opener-policy:bcg-cross-origin-isolated

Replace:
```
If navigationCOOP is "same-origin-plus-COEP", then set newBrowsingContext's group's cross-origin isolated to true.
```

By:
```
If navigationCOOP is "same-origin-plus-COEP", then set newBrowsingContext's group's cross-origin isolated to
- 'process-isolation' if canBeCrossOriginIsolated
- 'logical-isolation' otherwise.
```

What do you think of this in general?

annevk commented:
Yeah, I think that's right. And note that aside from the browsing context group's cross-origin isolated, you'll also need to do this for the similar boolean on agent cluster (which just copies the value from the browsing context group it's associated with).

ArthurSonzogni commented:
@annevk : The patch with your suggestions from https://github.com/whatwg/html/pull/6098#issuecomment-730270320 is ready.
If you have some time, could you please take a look?

annevk commented on source:
```diff
@@ -8078,7 +8078,8 @@ interface <dfn>DOMStringList</dfn> {
        <li>
-        <p>If <var>agentCluster</var>'s <span>cross-origin isolated</span> is false, then throw a
+        <p>If <var>agentCluster</var>'s <span>cross-origin isolated</span> is not <code
+        data-x="bcg-isolation-physic">isolation-physic</code>, then throw a
```
I'm surprised this isn't checking the capability. Wouldn't that mean that a cross-origin document can use `SharedArrayBuffer`?

cc @domenic @yutakahirano

ArthurSonzogni replied:
Now you say it, I am surprised too. My understanding is that we should check the capability.

That's what we implemented on Chrome here:
https://source.chromium.org/chromium/chromium/src/+/master:third_party/blink/renderer/core/execution_context/execution_context.cc;l=184;bpv=1;bpt=1?q=sharedarraybuffertran

ArthurSonzogni replied:
However, this requires talking about the destination realms.

I had some questions about it previously: https://crbug.com/1144838 
Here is a copy of my questions:
```
I discussed:
https://chromium-review.googlesource.com/c/chromium/src/+/2343658
with sigurds@. It made me ask myself several questions, that I answered by adding tests:
https://chromium-review.googlesource.com/c/chromium/src/+/2514211

I can now say the following:

From
- a COOP/COEP page,
- an OS without SharedArrayBuffer enabled by default (e.g. Android),
Then it is possible to share a SharedArrayBuffer from a non crossOriginIsolated document iframe.

Specifically, we have:
1. (sender.crossOriginIsolated == false) + (receiver.crossOriginIsolated == false) => Blocked.
2. (sender.crossOriginIsolated == true ) + (receiver.crossOriginIsolated == true ) => Transfered
3. (sender.crossOriginIsolated == true ) + (receiver.crossOriginIsolated == false) => Blocked
4. (sender.crossOriginIsolated == false) + (receiver.crossOriginIsolated == true ) => Transfered

This is not necessary bad, what matters is the receiver.
I wanted to make sure you are okay with case 4. WDYT?
```
I made some tests here:
https://chromium-review.googlesource.com/c/chromium/src/+/2514211
Unfortunately, not WPT test, because this rely on flags not enabled yet.

In the spec, I don't know the destination realms.
I guess I should update:
```
StructuredSerializeInternal ( value, forStorage [ , memory ] )
```
into
```
StructuredSerializeInternal ( value, forStorage [ , memory ] [, destinationRealm] )
```

and plumb |destinationRealm| everywhere this function is called.

annevk replied:
I think we should split this part out into its own PR given it's quite separate from what this PR is doing. Having said that, I think we can also check this when serializing as there is a current Realm. The difference here is whether `postMessage()` throws or the target gets a `messageerror` event. I suspect we want `postMessage()` to throw, though I also don't care too strongly. I was hoping @domenic and @yutakahirano would chime in.

domenic replied:
Yeah, woah, let's not make changes to structured serialization in this PR, especially one that changes the signature. That's a big change that needs more discussion in a dedicated issue, not a comment thread on a PR.

ArthurSonzogni replied:
I moved it out.

annevk commented on source:
```diff
@@ -78014,7 +78015,52 @@ console.assert(iframeWindow.frameElement === null);
 
   <p>A <span>browsing context group</span> has a <dfn data-x="bcg cross-origin
-  isolated">cross-origin isolated</dfn> boolean. It is initially false.</p>
+  isolated">cross-origin isolated</dfn> state. It can take the following values:</p>
```
I think we need to bikeshed these values a bit. Maybe "none", "logical", and "concrete"?

I don't think we need a separate value that combines the not-"none" values as we wouldn't be able to ever set it to that, right?

ArthurSonzogni replied:
I think this is better indeed. Thanks!

annevk commented on source:
```diff
@@ -78014,7 +78015,52 @@ console.assert(iframeWindow.frameElement === null);
+    <div class="note">
+    <p>In this group:</p>
+     <ul>
+      <li><p>Every <span>browsing context</span> are using
```
Browsing contexts don't hold headers. Documents or workers do.

annevk commented on source:
```diff
@@ -80543,8 +80594,9 @@ interface <dfn>BarProp</dfn> {
-  <span>agent cluster</span> are automatically origin-isolated. The `<code
+  <p class="note">Similarly, <code>Document</code>s with <span>agent cluster</span>'s
+  <span>cross-origin isolated</span> <code data-x="bcg-isolation-physic">isolation-physic</code>
+  are automatically origin-isolated. The `<code
```
I think all the not-"none" values offer the same features as Origin Isolation in terms of what is observable to web developers.

annevk commented on source:
```diff
@@ -80911,7 +80963,7 @@ interface <dfn>BarProp</dfn> {
     addition that it sets the (new) <span>top-level browsing context</span>'s <span data-x="tlbc
     group">group</span>'s <span data-x="bcg cross-origin isolated">cross-origin isolated</span> to
-    true.</p>
+    to <code data-x="bcg-isolation">isolation</code></p>
```
> a not-"none" value

seems more accurate.

annevk commented on source:
```diff
@@ -81079,7 +81131,8 @@ interface <dfn>BarProp</dfn> {
    <var>newBrowsingContext</var>'s <span data-x="tlbc group">group</span>'s <span data-x="bcg
-   cross-origin isolated">cross-origin isolated</span> to true.</p></li>
+   cross-origin isolated">cross-origin isolated</span> to <code
+   data-x="bcg-isolation">isolation</code>.</p></li>
```
I think we should call out the implementation-defined aspect and the two values here as this is where the "magic" happens.

annevk commented on source:
```diff
@@ -88313,8 +88367,10 @@ interface <dfn>ApplicationCache</dfn> : <span>EventTarget</span> {
-  <p>An <span>agent cluster</span> has an associated <dfn>cross-origin isolated</dfn> (a boolean),
-  which is initially false.</p>
+  <p>An <span>agent cluster</span> has an associated <dfn>cross-origin isolated</dfn> state. It
+  can take the same values as the <span>browsing context group</span>'s <span data-x="bcg
```
It might be cleanest to define "cross-origin-isolated value" and when introducing both slots you say they take such a value.

annevk reviewed (commented):
I think this looks good overall. My concerns are mainly about something that was already there (SAB protection) and the exact wording to use in the various places that need changes, but the architecture seems sound to me.

yutakahirano commented on source:
```diff
@@ -8087,14 +8087,16 @@ interface <dfn>DOMStringList</dfn> {
        <li>
-        <p>If <var>agentCluster</var>'s <span>cross-origin isolated</span> is false, then throw a
-        <span>"<code>DataCloneError</code>"</span> <code>DOMException</code>.</p>
+        <p>If <var>agentCluster</var>'s <span
```
agent' cluster's cross-origin isolation is not none?

domenic replied:
Yeah, capability is the property of a settings object, not an agent cluster. I think this change is what we want to work on in a separate bug.

domenic replied:
Hmm, or we can just change this to check the sender's capability. The separate bug can then be used to check whether we're going to check the destination.

yutakahirano commented on source:
```diff
@@ -99235,8 +99303,8 @@ interface <dfn>SharedWorkerGlobalScope</dfn> : <span>WorkerGlobalScope</span> {
-     capability</span> to <var>agent</var>'s <span>agent cluster</span>'s <span>cross-origin
-     isolated</span>.</p></li>
+     capability</span> to <var>agent</var>'s <span>agent cluster</span>'s <var
+     data-x="agent-cluster-cross-origin-isolation">cross-origin-isolation</var>.</p></li>
```
whether agent cluster's cross-origin isolation is concrete.

domenic commented on source:
```diff
@@ -77984,8 +77986,43 @@ console.assert(iframeWindow.frameElement === null);
-  <p>A <span>browsing context group</span> has a <dfn data-x="bcg cross-origin
-  isolated">cross-origin isolated</dfn> boolean. It is initially false.</p>
+  <p>A <span>browsing context group</span> has a <dfn><var
+  data-x="bcg-cross-origin-isolation">cross-origin-isolation</var></dfn> variable of type
```
This isn't a variable, so no "var".

Also, let's rename both this property and the below `<dfn>` to "cross-origin isolation mode"? Then this sentence would be

> A browsing context group has a cross-origin isolation mode, which is a cross-origin isolation mode.

domenic commented on source:
```diff
@@ -77984,8 +77986,43 @@ console.assert(iframeWindow.frameElement === null);
+
+  <p>A <dfn>cross-origin-isolation</dfn> type can take 3 possible values: </p>
+  <ul>
+   <li><dfn><code data-x="cross-origin-isolation-none">isolation-none</code></dfn></li>
```
Let's remove the isolation- prefix from each of these?

domenic reviewed (commented):
In penance for the long delays in review, I'll help out by pushing some commits to fix the issues I mention in this review, plus some editorial touchups.

domenic commented:
@annevk @yutakahirano @ArthurSonzogni PTAL. My changes are:

- Editorial fixups (including things like "cross-origin-isolation" -> "cross-origin isolation mode")
- Fixes to the issues @yutakahirano identified
- Changed the structured serialization to check the capability of the current realm. (We can discuss whether it checks the destination realm in a separate issue; IMO the spec is correct that such checking is not necessary, but we can discuss.)
- Since cross-origin isolated mode = logical/concrete => is origin-keyed = true (but not vice-versa), I updated two locations where it was more correct to talk about is origin-keyed. I.e. document.domain was checking both conditions, so I simplified it to only check is origin-keyed. And the non-normative note about window agent clusters was talking only about cross-origin isolated mode, but what it was saying was applicable to both, so I updated it to talk about origin-keying.

annevk commented on source:
```diff
@@ -80423,9 +80462,6 @@ interface <dfn>BarProp</dfn> {
    a <span>"<code>SecurityError</code>"</span> <code>DOMException</code>.</p></li>
 
-   <li><p>If the <span>surrounding agent</span>'s <span>agent cluster</span>'s <span>cross-origin
-   isolated</span> is true, then return.</p></li>
```
Why is this removed? Is it redundant with is origin-keyed now?

domenic replied:
Yes, exactly. Another thing to spell out in the commit message.

annevk commented on source:
```diff
@@ -81308,10 +81346,22 @@ interface <dfn>BarProp</dfn> {
+    <p class="note">It is difficult on some platforms to provide the security properties required by
+    the <span data-x="concept-settings-object-cross-origin-isolated-capability">cross-origin
+    isolated capability</span>. Only "<code
+    data-x="cross-origin-isolation-concrete">concrete</code>" might grant access to it. "<code
```
s/might/will/

ArthurSonzogni replied:
Using "might" was intentional. This also depends on the feature policy bit.

For instance: BCG `cross-origin-isolation-concrete` + document using `Feature-Policy: cross-origin-isolated 'none'` won't grant the `window.crossOriginIsolated` bit.

ArthurSonzogni replied:
In my last patchset. I used `can`. It sounds less confusing than `might`, and more precise than `will`.

annevk commented on source:
```diff
@@ -81308,10 +81346,22 @@ interface <dfn>BarProp</dfn> {
+    isolated capability</span>. Only "<code
+    data-x="cross-origin-isolation-concrete">concrete</code>" might grant access to it. "<code
+    data-x="cross-origin-isolation-logical">logical</code>" won't, and is used by implementations on
+    other platforms.</p>
```
This other is weird, since we already started with some platforms that might use "logical".

ArthurSonzogni replied:
I agree. I removed `,and is used by implementations on other platforms`.

annevk reviewed (approved):
I was thinking that we would split out the structured serialization bug fix entirely, but I guess it's okay to include it here if we spell it out in the commit message. I found a couple nits, seems okay to me to land this modulo those, but I'd appreciate one more review since it's a bit harder to review on a small screen.

ArthurSonzogni reviewed (commented):
Thanks @domenic for the big update!

annevk commented:
@ArthurSonzogni could you prepare a tentative commit message that also accounts for the additional changes as discussed above and post it as a comment? I think this is good to go. (I pushed one more nit.)

ArthurSonzogni commented:
Thanks @annevk !

Tentative commit message:
```git
The [specification] currently requires [COOP] + [COEP] to give access to
crossOriginIsolated capabilities like SharedArrayBuffer.

Some platforms can't easily support multiple processes (like Android
Webview). Therefore, they can't really support crossOriginIsolated.
However the are no strong reasons for them not to enforce COEP (and
maybe COOP) when their associated headers are present.

This patch makes the specification to allow (instead of requiring)
platform to grant the crossOriginIsolated capability when both COOP and
COEP are used.

The browsing context group's cross-origin-isolation mode becomes a
tri-state:
- none     (not COOP+COEP)
- logical  (COOP+COEP, without granting crossOriginIsolated)
- concrete (COOP+COEP, with granting crossOriginIsolated)

[Bug]: https://github.com/whatwg/html/issues/6060

[specification]:
https://html.spec.whatwg.org/#cross-origin-opener-policies

[COOP]:
https://html.spec.whatwg.org/#cross-origin-opener-policy

[COEP]:
https://html.spec.whatwg.org/#coep

---

Additional minor change:

When setting document.domain, both properties were previously checked:
1) agent-cluster's cross-origin-isolation mode.
2) agent-cluster's origin-keyed.
The first has been removed, since it already implies the second.
```

annevk commented:
Thanks, we should also mention the change to StructuredSerializeInternal, but I can merge that in as well. If there are no further comments I'll merge this tomorrow.

ArthurSonzogni commented:
> Thanks, we should also mention the change to StructuredSerializeInternal, but I can merge that in as well. If there are no further comments I'll merge this tomorrow.

Did we changed StructuredSerializeInternal? I see only some "data-x=" being now properly set. The content is the same.
We previously discussed updating StructuredSerializeInternal, but we said this was unrelated to this current patch, so I removed the change.

annevk commented:
It went from checking cross-origin isolated to cross-origin isolated capability, right? That's a big change, even if one could consider it an oversight from when we introduced cross-origin isolated capability.

ArthurSonzogni commented:
Indeed! Updated tentative commit message. (Feel free to improve it, if needed)

```git
Make COOP+COEP do not imply crossOriginIsolated.

This contains 3 changes:
---

The [specification] currently requires [COOP] + [COEP] to give access to
crossOriginIsolated capabilities like SharedArrayBuffer.

Some platforms can't easily support multiple processes (like Android
Webview). Therefore, they can't really support crossOriginIsolated.
However the are no strong reasons for them not to enforce COEP (and
maybe COOP) when their associated headers are present.

This patch makes the specification to allow (instead of requiring)
platform to grant the crossOriginIsolated capability when both COOP and
COEP are used.

The browsing context group's cross-origin-isolation mode becomes a
tri-state:
- none     (not COOP+COEP)
- logical  (COOP+COEP, without granting crossOriginIsolated)
- concrete (COOP+COEP, with granting crossOriginIsolated)

[Bug]:  https://github.com/whatwg/html/issues/6060
[COOP]: https://html.spec.whatwg.org/#cross-origin-opener-policy
[COEP]: https://html.spec.whatwg.org/#coep
[specification]:
https://html.spec.whatwg.org/#cross-origin-opener-policies

---

When setting document.domain, both properties were previously checked:
1. agent-cluster's cross-origin-isolation mode.
2. agent-cluster's origin-keyed.
The first has been removed, since it already implies the second.

---

When serializing a sharedArrayBuffer, check the cross-origin isolated
capability instead of the agent-cluster cross-origin isolation mode.
This was an oversight from when the cross-origin isolated capability was
first introduced.
```

annevk commented:
Thanks so much @ArthurSonzogni!

ArthurSonzogni commented:
Thanks!
